src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java

```diff
@@ -4,57 +4,72 @@
  */
 package org.geoserver.ows;
 
-import java.util.Collections;
-import java.util.Enumeration;
+import java.util.Arrays;
 import java.util.Map;
-import javax.servlet.http.HttpServletRequest;
+import java.util.Optional;
+import java.util.function.Predicate;
 
-/** Mangles service URL's based on the AcceptLanguages and Languages parameters */
+/**
+ * Mangles service for i18n AcceptLanguages and Languages parameters.
+ *
+ * <p>Based on the documentation
+ * https://docs.geoserver.org/latest/en/user/services/internationalization/index.html this class
+ * provide the capability to handle the AcceptLanguages and also the custom INSPIRE extension
+ * Language parameter
+ * (https://docs.geoserver.org/stable/en/user/extensions/inspire/using.html#internationalization-support)
+ * for support capabilities in multiple languages
+ */
 public class LanguageURLMangler implements URLMangler {
 
     public static final String ACCEPT_LANGUAGES = "AcceptLanguages";
     public static final String LANGUAGE = "Language";
 
+    private final Predicate<Object> filterEmtpyString = language -> !((String) language).isEmpty();
+
     @Override
     public void mangleURL(
             StringBuilder baseURL, StringBuilder path, Map<String, String> kvp, URLType type) {
-        if (Dispatcher.REQUEST.get() == null || !type.equals(URLType.SERVICE)) return;
-
-        String languageParameter = null;
-        String acceptLanguagesParameter = null;
-
-        HttpServletRequest httpRequest = Dispatcher.REQUEST.get().getHttpRequest();
-        Enumeration<String> parameterNames = httpRequest.getParameterNames();
-
-        for (String parameter : Collections.list(parameterNames)) {
-            if (parameter.equalsIgnoreCase(LANGUAGE)) {
-                languageParameter = parameter;
-            } else if (parameter.equalsIgnoreCase(ACCEPT_LANGUAGES)) {
-                acceptLanguagesParameter = parameter;
-            }
-        }
-
-        String language = "";
-
-        if (acceptLanguagesParameter != null) {
-            String acceptLanguages = httpRequest.getParameter(acceptLanguagesParameter);
-            if (acceptLanguages != null) {
-                String commaSplit = acceptLanguages.split(",")[0];
-                String spaceSplit = commaSplit.split(" ")[0];
-                if (spaceSplit.length() >= 1) {
-                    language = spaceSplit;
-                } else {
-                    language = commaSplit;
-                }
-            }
-        }
-
-        if (languageParameter != null) {
-            language = httpRequest.getParameter(languageParameter);
-        }
-
-        if (language != null && language.length() > 0) {
-            kvp.put("Language", language.trim());
-        }
+        Optional.ofNullable(type)
```

Comment:
Careful about this proliferation of Optional calls, as they come with object allocations, one for the main Optional, plus others for specific methods (e.g. map, flatmap).

In this code, it's probably not harmful, a single request processing normally builds up to a few tens of URLs, so overall the code change introduces the allocation of the order of 100 (ish) new objects that weren't there in the old code. In the economy of a OGC request, that's peanuts.

However, always consider the context in which the code is running, if it's called many times (hundreds of thousands, millions, not unusual when dealing with GIS data) best to stay away from optionals and streams due to their memory and execution costs.

Reply:
I perfectly got your point, and here the `type` parameter wrapped in an Optional is for sure useless. I will revert it with the if.